packages/steps/src/BuildFunction.ts

```diff
@@ -1,78 +1,96 @@
+import { BuildStep, BuildStepFunction } from './BuildStep.js';
 import { BuildStepContext } from './BuildStepContext.js';
-import { BuildStepInputCreator } from './BuildStepInput.js';
-import { BuildStepOutputCreator } from './BuildStepOutput.js';
+import { BuildStepInputProvider } from './BuildStepInput.js';
+import { BuildStepOutputProvider } from './BuildStepOutput.js';
 
 export type BuildFunctionById = Record<string, BuildFunction>;
 export type BuildFunctionCallInputs = Record<string, string>;
 
 export class BuildFunction {
-  public readonly id?: string;
+  public readonly namespace?: string;
+  public readonly id: string;
   public readonly name?: string;
   public readonly platforms?: BuildPlatform[];
-  public readonly inputCreators?: BuildStepInputCreator[];
-  public readonly outputCreators?: BuildStepOutputCreator[];
-  public readonly command: string;
+  public readonly inputProviders?: BuildStepInputProvider[];
+  public readonly outputProviders?: BuildStepOutputProvider[];
+  public readonly command?: string;
+  public readonly fn?: BuildStepFunction;
   public readonly shell?: string;
 
-  constructor(
-    private readonly ctx: BuildStepContext,
-    {
-      id,
-      name,
-      platforms,
-      inputCreators,
-      outputCreators,
-      command,
-      shell,
-    }: {
-      id?: string;
-      name?: string;
-      platforms?: BuildPlatform[];
-      inputCreators?: BuildStepInputCreator[];
-      outputCreators?: BuildStepOutputCreator[];
-      command: string;
-      shell?: string;
-    }
-  ) {
+  constructor({
+    namespace,
+    id,
+    name,
+    platforms,
+    inputProviders,
+    outputProviders,
+    command,
+    fn,
+    shell,
+  }: {
+    namespace?: string;
+    id: string;
+    name?: string;
+    platforms?: BuildPlatform[];
+    inputProviders?: BuildStepInputProvider[];
+    outputProviders?: BuildStepOutputProvider[];
+    command?: string;
+    fn?: BuildStepFunction;
+    shell?: string;
+  }) {
+    assert(command !== undefined || fn !== undefined, 'Either command or fn must be defined.');
+    assert(!(command !== undefined && fn !== undefined), 'Command and fn cannot be both set.');
+
+    this.namespace = namespace;
     this.id = id;
     this.name = name;
     this.platforms = platforms;
-    this.inputCreators = inputCreators;
-    this.outputCreators = outputCreators;
+    this.inputProviders = inputProviders;
+    this.outputProviders = outputProviders;
     this.command = command;
+    this.fn = fn;
     this.shell = shell;
   }
 
-  public createBuildStepFromFunctionCall({
-    id,
-    callInputs = {},
-    workingDirectory,
-    shell,
-  }: {
-    id?: string;
-    callInputs?: BuildFunctionCallInputs;
-    workingDirectory: string;
-    shell?: string;
-  }): BuildStep {
-    const buildStepId = id ?? this.id ?? uuidv4();
+  public getFullId(): string {
+    return this.namespace === undefined ? this.id : `${this.namespace}/${this.id}`;
```

Comment:
If we are using `/` as namespace separator we should limit what can be used for regular ids. I would suggest limiting it to `^(?:eas\/)?[a-zA-Z0-9_-]*$` (not tested).